File: 152-maximum-product-subarray/152-maximum-product-subarray.py
class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        maxi = nums[0]
        mini = nums[0]
        result = nums[0]
        for i in range(1, len(nums)):
            temp = maxi
            maxi = max(maxi*nums[i], mini*nums[i], nums[i])
            mini = min(temp * nums[i], mini * nums[i], nums[i])
            print(maxi)
            if maxi > result:  result = maxi
        return result
        

File: 152-maximum-product-subarray/152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int max = nums[0], min = nums[0];
        int result = nums[0];
        for (int i=1; i<nums.length; i++) {
            int temp = min;
            min = Math.min(Math.min(nums[i]*min, nums[i]*max), nums[i]);
            max = Math.max(Math.max(nums[i]*temp, nums[i]*max), nums[i]);
            result = Math.max(max, result);
        }
        return result;
    }
}